File: src/rss.ts
import RSS from 'rss';
import { MeetupEvent } from './types';

export function generateRSSFeed(
  events: MeetupEvent[],
  meetupUrl: string,
  feedType: 'upcoming' | 'past'
): string {
  const groupName = extractGroupName(meetupUrl);

  const feed = new RSS({
    title: `${groupName} - ${feedType === 'upcoming' ? 'Upcoming' : 'Past'} Events`,
    description: `${feedType === 'upcoming' ? 'Upcoming' : 'Past'} meetup events for ${groupName}`,
    feed_url: `${meetupUrl}/feed/${feedType}`,
    site_url: meetupUrl,
    language: 'en',
    pubDate: new Date().toISOString(),
  });

  events.forEach((event) => {
    const locationStr = event.location
      ? `${event.location.name || ''} ${event.location.address || ''} ${event.location.city || ''}, ${event.location.state || ''}`.trim()
      : 'Location TBD';

    feed.item({
      title: event.title,
      description: `${event.description}\n\nWhen: ${new Date(event.dateTime).toLocaleString()}\nWhere: ${locationStr}\nRSVPs: ${event.going}`,
      url: event.eventUrl,
      date: event.dateTime,
      guid: event.id,
    });
  });

  return feed.xml();
}

function extractGroupName(url: string): string {
  const match = url.match(/meetup\.com\/([^/]+)/);
  return match ? match[1] : 'Meetup Group';
}

File: cmd/server/rss.go
package main

import (
	"time"

	"github.com/gorilla/feeds"
)

func GenerateRSSFeed(events []MeetupEvent, meetupURL, feedType string) (string, error) {
	feed := &feeds.Feed{
		Title:       "Meetup Events - " + feedType,
		Link:        &feeds.Link{Href: meetupURL},
		Description: "Meetup.com events feed",
		Created:     time.Now(),
	}

	for _, event := range events {
		item := &feeds.Item{
			Title:       event.Title,
			Link:        &feeds.Link{Href: event.EventURL},
			Description: event.Description,
			Created:     time.Now(),
		}

		// Parse dateTime if available
		if event.DateTime != "" {
			if t, err := time.Parse(time.RFC3339, event.DateTime); err == nil {
				item.Created = t
			}
		}

		// Add location to description if available
		if event.Location != nil {
			loc := ""
			if event.Location.Name != "" {
				loc += event.Location.Name
			}
			if event.Location.City != "" {
				if loc != "" {
					loc += ", "
				}
				loc += event.Location.City
			}
			if event.Location.State != "" {
				if loc != "" {
					loc += ", "
				}
				loc += event.Location.State
			}
			if loc != "" {
				item.Description += "\n\nLocation: " + loc
			}
		}

		feed.Items = append(feed.Items, item)
	}

	return feed.ToRss()
}
